### PR TITLE
reorganize sort headers

### PR DESCRIPTION
**Pull request by fnrizzi**
Reorganizes the sorting header files for readability and separation of public/impl.

### Before: 
```
algorithms/
├── CMakeLists.txt
├── src
│   ├── CMakeLists.txt
│   ├── KokkosAlgorithms_dummy.cpp
│   ├── Kokkos_NestedSort.hpp
│   ├── Kokkos_Random.hpp
│   ├── Kokkos_Sort.hpp
│   ├── Kokkos_StdAlgorithms.hpp
```
Issues with this: `Kokkos_Sort.hpp` contains *all* the public APIs and *impl* for sort and bin sort, making it hard to follow and understand what is going on, and exposes all the impl details too. The same is happening for `Kokkos_NestedSort.hpp`

### This PR addresses the above as follows
```
algorithms/
├── CMakeLists.txt
├── src
│   ├── CMakeLists.txt
│   ├── KokkosAlgorithms_dummy.cpp
│   ├── Kokkos_NestedSort.hpp
│   ├── Kokkos_Random.hpp
│   ├── Kokkos_Sort.hpp
│   ├── Kokkos_StdAlgorithms.hpp
│   ├── sorting
│   │   ├── Kokkos_BinOpsPublicAPI.hpp
│   │   ├── Kokkos_BinSortPublicAPI.hpp
│   │   ├── Kokkos_NestedSortPublicAPI.hpp
│   │   ├── Kokkos_SortPublicAPI.hpp
│   │   └── impl
│   │       ├── Kokkos_CopyOpsForBinSortImpl.hpp
│   │       ├── Kokkos_NestedSortImpl.hpp
│   │       └── Kokkos_SortImpl.hpp
│   └── std_algorithms
```
- Public API and impl details are separated, therefore not exposing impl details to users
- `Kokkos_Sort.hpp` remains the header to include to access *all* sorting functionalities. 

Related to #6196

### CODE REVIEW

File: algorithms/src/sorting/impl/Kokkos_NestedSort.hpp
Comment (masterleinad) [inline algorithms/src/sorting/impl/Kokkos_NestedSort.hpp:1]: Maybe, call it `Kokkos_NestedRange.hpp`?

File: algorithms/src/Kokkos_NestedSort.hpp [hunk 1]
```diff
@@ -14,175 +14,17 @@
...
-template <class TeamMember, class KeyViewType, class ValueViewType,
-          class Comparator>
-KOKKOS_INLINE_FUNCTION void sort_by_key_thread(const TeamMember& t,
-                                               const KeyViewType& keyView,
-                                               const ValueViewType& valueView,
-                                               const Comparator& comp) {
-  Impl::sort_nested_impl(t, keyView, valueView, comp,
-                         Impl::NestedRange<false>());
-}
+#include "sorting/Kokkos_NestedSort.hpp"
```
Comment (masterleinad) [inline algorithms/src/Kokkos_NestedSort.hpp:24]: Can't this just be `sorting/Kokkos_NestedSort.hpp` avoiding the indirection?

File: algorithms/src/sorting/Kokkos_Sort.hpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+  }
+  if (std::is_floating_point<typename ViewType::non_const_value_type>::value) {
+    KOKKOS_ASSERT(std::isfinite(static_cast<double>(result.max_val) -
+                                static_cast<double>(result.min_val)));
+  }
+
+  BinSort<ViewType, CompType> bin_sort(
+      view, CompType(max_bins, result.min_val, result.max_val), sort_in_bins);
+  bin_sort.create_permute_vector(exec);
+  bin_sort.sort(exec, view);
```
Comment (masterleinad) [inline algorithms/src/sorting/Kokkos_Sort.hpp:135]: This header needs `BinSort` but doesn't include it.

File: algorithms/src/sorting/Kokkos_Sort.hpp
Comment (masterleinad) [inline algorithms/src/sorting/Kokkos_Sort.hpp:1]: Maybe, `Kokkos_SortFreeFunctions.hpp`?

**Review by masterleinad — COMMENTED** [review]
I find it confusing to have the name header file name just in different nested folders. Also, is it feasible to make the individual header-files standalone?

**Comment by fnrizzi** [comment]
@masterleinad how about : 

```
│   ├── Kokkos_Random.hpp
│   ├── Kokkos_NestedSort.hpp
│   ├── Kokkos_Sort.hpp
│   ├── Kokkos_StdAlgorithms.hpp
│   ├── sorting
│   │   ├── Kokkos_BinOpsPublicApi.hpp
│   │   ├── Kokkos_BinSortPublicApi.hpp
│   │   ├── Kokkos_NestedSortPublicApi.hpp
│   │   ├── Kokkos_SortPublicApi.hpp
│   │   └── impl
│   │       ├── Kokkos_BinSortImpl.hpp
│   │       ├── Kokkos_NestedSortImpl.hpp
```

**Comment by fnrizzi** [comment]
> Also, is it feasible to make the individual header-files standalone?

all of them? by standalone you mean they include all the necessary things ?

**Comment by masterleinad** [comment]
> by standalone you mean they include all the necessary things ?

Yes

File: algorithms/src/sorting/Kokkos_Sort.hpp
Comment (fnrizzi) [reply algorithms/src/sorting/Kokkos_Sort.hpp:1]: yes that is what I had in the other PR inside the impl

File: algorithms/src/sorting/Kokkos_Sort.hpp (hunk 2 above)
Comment (fnrizzi) [reply algorithms/src/sorting/Kokkos_Sort.hpp:135]: adding it thanks

**Comment by masterleinad** [comment]
> @masterleinad how about?

That looks reasonable to me.

File: algorithms/src/Kokkos_NestedSort.hpp (hunk 1 above)
Comment (fnrizzi) [reply algorithms/src/Kokkos_NestedSort.hpp]: sorry don't understand, what do you mean?

**Comment by fnrizzi** [comment]
@masterleinad i think i addressed your comments

**Review by masterleinad — APPROVED** [review]
FIne with me.

**Comment by fnrizzi** [comment]
Thanks for the review Daniel!

**Review by dalg24 — APPROVED** [review]

**Pull request merged**